### ISSUE TITLE
MidiPlayer.cs not thread-safe, crashes in Mute()

### ISSUE THREAD

**Issue by jeroenwalter** [atsushieno/managed-midi#64]
When a midi file is playing and you call Stop or Pause or call the Seek method, the Mute() method is called from another thread than the play thread.
This will cause the MidiPlayer.buffer field to be written in 2 threads at once.
This will cause an exception in MidiEvent.Convert, as it can't convert the message correctly, as the contents of the message is changed during conversion.

I can reproduce this quite easily. I have a program that has a scrollbar for setting the playback position via Seek(). Playing the song and then using the scrollbar to repeatedly call Seek() will almost immediately trigger the crash.

A solution can be by not having the buffer field in MidiPlayer, but using a local array in the eventhandlers for MidiPlayer.EventReceived and MidiPlayer.player.Starting in MidiPlayer.cs

**Comment by jeroenwalter** [comment]
Also, for Stop and Pause, it would be better to actually wait until playback has stopped, before calling Mute, otherwise the midi messages from Mute can be played before the playback is stopped. If any midi notes are then played by the player thread, you will hear them.

**Comment by jeroenwalter** [comment]
Better yet, make sure only the player thread can send midi messages.
This means that Pause and Seek should not call Mute. The same goes for Dispose.
Also, Pause, Stop and Seek should become awaitable Tasks, and Dispose must only dispose the output once playback has completely stopped, otherwise midi messages can be sent to a disposed output.

**Comment by jeroenwalter** [comment]
It's too bad there are so few unit tests for this project, otherwise I would immediately tried fixing this.
Now I have to put more effort in trying to understand the code and writing unit tests that prove that there is a bug, and then fixing it.

**Comment by atsushieno** [comment]
Hi, thanks for detailed insights on the current MidiPlayer implementation. I welcome patches but don't plan to actively working on this project by myself unless they are relevant to my related projects, as my [gist](https://gist.github.com/atsushieno/2830490dabfe37190667d8a31ef562f2) on my profile page states. My gut feeling is that working on .NET stuff is time-wasting to me until .NET 5 or any later versions gets released and I don't have to be messed by those TFM differences and I have many other fun projects to work on to waste time on development platform that does not support me / my environment (Linux desktop).

There have been always little chance to get platform-dependent tests in comfortable experience, which is the most of the target of this project.

**Comment by jeroenwalter** [comment]
No problem, your managed-midi classes are very helpful and I want to thank you for that.
For my project they need to be rocksteady, so any bugs I find, I intend to fix.

I've already started on fixing some things, starting with some unit tests.
When I've got something running, I may create a pull request.
I'm working on a project using a Raspberry Pi and netcore 3.1 and fluidsynth (so I will likely also take a look at your nfluidsynth). 
Development for now is in Windows, but I've already had a prototype working on the rpi a few months ago.

**Comment by jeroenwalter** [comment]
I've started work on squashing this bug in my own fork.
You can follow it on https://github.com/jeroenwalter/managed-midi/tree/threadsafety
I've mainly done some refactoring work, moved classes to their own files etc.
I've also added a new WpfMidiPlayer project, so I can more easily reproduce and debug the bug.

**Comment by jeroenwalter** [comment]
I've fixed the bug in my fork, but I don't know if it's wise to send a pull request, as I have made some more changes and I'm not yet happy with the state the code is in now. 
Also I found some more bugs and I'm considering refactoring the timing mechanism so it doesn't block the thread in WaitBy(), as that has some drawbacks for my application and it affect how Seek works.

The actual fix for the crash is not difficult, just remove the buffer field from the class and allocate the buffer for each midi event. 
See this commit: https://github.com/jeroenwalter/managed-midi/commit/45e383272338c46618e05f014987efbadd65a725

**Comment by atsushieno** [comment]
Awesome! Thanks for letting me know.

Let's take best of your productivity. Which scenario is the best for you?

- Your changes  are mostly acceptable, I can have those split sources as well as your Wpf sample in my master branch. Although I cannot take care of the Wpf app as it does not run on my Linux desktop environment, so it will be left unsupported.
  - As an exceptional scenario, I can give you a commit access to the repo. I don't make changes these days anymore, but there can also be other PRs that I would accept.
- You continue development on your new master (which likely means, a new development head), if you plan to fix further issues. I can steal your changes as long as it is based on the same license back to my master. I can send future PRs to your tree too if  they make sense (unless things are PR-ed to my tree).

**Comment by jeroenwalter** [comment]
I don't think commit access is necessary.

To give you some context of what I'm developing:
I'm creating the next version of my electronic barrel organ application ([https://www.draaimuziek.nl/](https://www.draaimuziek.nl/)). 
I've already developed a C++ Windows application several years ago and now I want create a C# version running on the Raspberry Pi, that is controllable via a web app on a smartphone.
It uses an Arduino to register the speed of the spinning wheel and uses it to change the tempo of the midi music on the fly.

Right now I'm focussing on getting my application in a somewhat workable prototyping state.
The application will eventually consists of the following: the midi player, a database using Entity Framework Core, input from an Arduino, a REST/ODATA interface (asp.net core 5) and a web app in vue.js (or maybe Blazor WebAssembly). 
The application must work on a Raspberry Pi with fluidsynth.
This is all a lot of work and I'm doing this in my free time.

I'm trying to focus on the most important part, i.e. the midi player and controlling its tempo via an Arduino.
And here is where I am now: looking at the midi player and the MidiEventLooper and how it controls the timing via SimpleAdjustingMidiPlayerTimeManager.WaitBy. 
The current mechanism doesn't allow for a tempo of 0.0 or for fast tempo changes.

In order to make it workable for me, I need to rewrite the MidiEventLooper so it uses a 1ms timer instead of a Task.Delay.
That's quite a big change and I need that to be tested for performance on a Raspberry Pi.


I'm reluctant to make breaking changes, as that would make it harder for you to use my changes in your master branch.
If I want to avoid making breaking changes, I think I just need to create new MidiPlayer and MidiEventLooper classes.
That's fine by me, you can then later on decide if you are pleased with it and incorporate them in your master branch.

I also have some issues with some other aspects of your API, but I also want to avoid breaking changes here.
Mostly they are to do with asynchronous methods that I think should be synchronous, such as start, stop and seek. Also I'm missing logging functionality.
You've marked some classes and methods obsolete with the intention of creating a new API, but I don't see that happening anytime soon.
I quite like the current MidiAccessManager and related classes, they are easy to use, so maybe you can detail what you intend to do here and if you still are planning to do so (and when).

So in short.... I prefer to develop my stuff first and in the meantime if you can use some, just cherry pick it for your master branch and then I'll sync my branch with it again later on.
Or when I'm finished, we can do one big pull request :)
